pyClient/zeth/contracts.py

```diff
@@ -4,40 +4,77 @@
 #
 # SPDX-License-Identifier: LGPL-3.0+
 
+from __future__ import annotations
 import zeth.constants as constants
 from zeth.encryption import EncryptionPublicKey, encode_encryption_public_key
 from zeth.signing import SigningVerificationKey
-from zeth.zksnark import IZKSnarkProvider, GenericProof
-from zeth.utils import get_trusted_setup_dir, get_contracts_dir, hex_to_int, \
-    get_public_key_from_bytes
+from zeth.zksnark import IZKSnarkProvider, GenericProof, GenericVerificationKey
+from zeth.utils import get_contracts_dir, hex_to_int, get_public_key_from_bytes
 
-import json
 import os
-from web3 import Web3, HTTPProvider  # type: ignore
-from solcx import compile_files  # type: ignore
-from typing import Tuple, Dict, List, Any
+from web3 import Web3  # type: ignore
+from solcx import compile_files      # type: ignore
+from typing import Tuple, Dict, List, Iterator, Optional, Any
 
-W3 = Web3(HTTPProvider(constants.WEB3_HTTP_PROVIDER))
-# pylint is not aware of W3.eth. This prevents it from complaining.
-eth = W3.eth  # pylint: disable=no-member, invalid-name
+# Avoid trying to read too much data into memory
+SYNC_BLOCKS_PER_BATCH = 10
 
 Interface = Dict[str, Any]
 
+# List of address and cipher text pairs
+EncryptedNote = Tuple[int, bytes]
+
 
 class MixResult:
     """
     Data structure representing the result of the mix call.
     """
     def __init__(
             self,
-            encrypted_notes: List[Tuple[int, bytes]],
+            encrypted_notes: List[EncryptedNote],
             new_merkle_root: str,
             sender_k_pk: EncryptionPublicKey):
         self.encrypted_notes = encrypted_notes
         self.new_merkle_root = new_merkle_root
         self.sender_k_pk = sender_k_pk
 
 
+class InstanceDescription:
+    """
+    Minimal data required to instantiate the in-memory interface to a contract.
+    """
+    def __init__(self, address: str, abi: Dict[str, Any]):
+        self.address = address
+        self.abi = abi
+
+    def to_json_dict(self) -> Dict[str, Any]:
+        return {
+            "address": self.address,
+            "abi": self.abi
+        }
+
+    @staticmethod
+    def from_json_dict(desc_json: Dict[str, Any]) -> InstanceDescription:
+        return InstanceDescription(desc_json["address"], desc_json["abi"])
+
+    @staticmethod
+    def from_instance(instance: Any) -> InstanceDescription:
+        """
+        Return the description of an existing deployed contract.
+        """
+        return InstanceDescription(instance.address, instance.abi)
+
+    def instantiate(self, web3: Any) -> Any:
+        """
+        Return the instantiated contract
+        """
+        return web3.eth.contract(address=self.address, abi=self.abi)
+
+
+def get_block_number(w3: Any) -> int:
```

Comment:
let's use either `w3` everywhere or stick to `web3` to be consistent throughout the code

Reply:
changed everywhere